trace_decoder/src/type1.rs

```diff
@@ -32,27 +30,140 @@ pub fn frontend(instructions: impl IntoIterator<Item = Instruction>) -> anyhow::
     );
     let execution = executions.into_vec().remove(0);
 
-    let mut visitor = Visitor {
-        path: Vec::new(),
-        frontend: Frontend {
-            state: HashedPartialTrie::default(),
-            code: HashSet::new(),
-            storage: HashMap::new(),
+    let mut frontend = Frontend::default();
+    visit(
+        &mut frontend,
+        &stackstack::Stack::new(),
+        match execution {
+            Execution::Leaf(it) => Node::Leaf(it),
+            Execution::Extension(it) => Node::Extension(it),
+            Execution::Branch(it) => Node::Branch(it),
+            Execution::Empty => Node::Empty,
         },
-    };
-    visitor.visit_node(match execution {
-        Execution::Leaf(it) => Node::Leaf(it),
-        Execution::Extension(it) => Node::Extension(it),
-        Execution::Branch(it) => Node::Branch(it),
-        Execution::Empty => Node::Empty,
-    })?;
-    let Visitor { path, frontend } = visitor;
-
-    assert_eq!(Vec::<U4>::new(), path);
+    )?;
 
     Ok(frontend)
 }
 
+fn visit(
+    frontend: &mut Frontend,
+    path: &stackstack::Stack<'_, U4>,
+    node: Node,
+) -> anyhow::Result<()> {
+    match node {
+        Node::Hash(Hash { raw_hash }) => {
+            frontend
+                .state
+                .insert_hash_by_path(TriePath::new(path.iter().copied())?, raw_hash.into())?;
+        }
+        Node::Leaf(Leaf { key, value }) => {
+            let path = TriePath::new(path.iter().copied().chain(key))?;
+            match value {
+                // TODO(0xaatif): what should this be interpreted as?
```

Comment:
Todo in this PR? Worth opening an issue?

Reply:
Left as an error - if it's a logic error we can come back to it